src/ui/reducers/sources.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  createEntityAdapter,
+  createSelector,
+  createSlice,
+  EntityState,
+  PayloadAction,
+} from "@reduxjs/toolkit";
+import { newSource, SourceKind } from "@replayio/protocol";
+import { getSelectedSourceId } from "devtools/client/debugger/src/selectors";
+import { UIState } from "ui/state";
+import { newSourcesToCompleteSourceDetails } from "ui/utils/sources";
+
+export interface SourceDetails {
+  canonicalId: string;
+  contentHash: string | undefined;
```

Comment:
:hand: All these `| undefined` fields could probably be `field?: whatever` instead